File: MergeSort.rb
def sort(list)
  len = list.size

  if len === 1
    return list
  end

  left, right = list.each_slice((len/2.0).round ).to_a

  left = sort(left)
  right = sort(right)

  return merge(left, right)

end

def merge(left, right)
  sorted = []
  while (!left.empty? && !right.empty?)
    if (left[0] > right[0])
      sorted.push(right[0])
      right.shift()
    else
      sorted.push(left[0])
      left.shift()
    end
  end

  while (!left.empty?)
    sorted.push(left[0])
    left.shift()
  end

  while (!right.empty?)
    sorted.push(right[0])
    right.shift()
  end
  sorted
end

sort([3,5,2,6,1])

File: MergeSort.js
function sort(arr) {
  let length = arr.length,
  mid = Math.floor(arr.length/2),
  left = arr.slice(0, mid),
  right = arr.slice(mid, length)
  if (length === 1) {
    return arr;
  }

  return merge(sort(left), sort(right))
}

function merge(left, right) {
  sorted = [];

  while(left.length !== 0 && right.length !== 0) {
    if (left[0] > right[0]) {
      sorted.push(right[0]);
      right.splice(0, 1);
    } else {
      sorted.push(left[0]);
      left.splice(0, 1);
    }
  }

  while (left.length !== 0) {
    sorted.push(left[0]);
    left.splice(left.length - 1, 1);
  }

  while (right.length !== 0) {
    sorted.push(right[0]);
    right.splice(right.length - 1, 1);
  }

  return sorted;
}
sort([2, 3, 6, 1])
